src/AppMonad.hs

```diff
@@ -0,0 +1,90 @@
+{-# LANGUAGE EmptyDataDecls             #-}
+{-# LANGUAGE FlexibleContexts           #-}
+{-# LANGUAGE FlexibleInstances          #-}
+{-# LANGUAGE GADTs                      #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE MultiParamTypeClasses      #-}
+{-# LANGUAGE OverloadedStrings          #-}
+{-# LANGUAGE QuasiQuotes                #-}
+{-# LANGUAGE RankNTypes                 #-}
+{-# LANGUAGE StandaloneDeriving         #-}
+{-# LANGUAGE TemplateHaskell            #-}
+{-# LANGUAGE TypeFamilies               #-}
+
+module AppMonad where
+
+import Control.Monad.Catch     (MonadCatch, MonadThrow)
+import Control.Monad.IO.Class  (MonadIO, liftIO)
+import Control.Monad.IO.Unlift (MonadUnliftIO)
+import Control.Monad.Logger    (NoLoggingT, runStdoutLoggingT)
+import Control.Monad.Reader    (MonadReader, ReaderT(ReaderT), asks, runReaderT)
+import Control.Monad.Trans     (lift)
+
+import Conduit
+import Data.Time
+import Database.Persist.Postgresql
+import Database.Persist.TH
+
+type App = AppT IO
+
+data Env =
+  Env
+    { connStr :: ConnectionString
+    , port    :: Int
+    }
+  deriving (Eq, Show)
+
+newtype AppT m a =
+  AppT (ReaderT Env m a)
+  deriving (Functor, Applicative, Monad, MonadIO, MonadThrow, MonadCatch, MonadReader Env)
+
+deriving instance MonadTrans AppT
+
+deriving instance MonadUnliftIO App
+
+share
+  [mkPersist sqlSettings, mkMigrate "migrateAll"]
+  [persistLowerCase|
+Player
+    name String
+    email String
+    deriving Show
+Item
+    symbol String
+    amount Int
+    price Double
+    idPlayer PlayerId
+    ts UTCTime
+    deriving Show
+|]
+
+type QueryT b a = ReaderT b (NoLoggingT (ResourceT IO)) a
+
+withDb ::
+     forall m b a. (MonadUnliftIO m, IsPersistBackend b, BaseBackend b ~ SqlBackend)
+  => QueryT b a
+  -> AppT m a
+withDb action = do
```

Comment:
Ideally the signature of `withDb` should be something like `(T1 m, T2 m..) => App a -> App a` or `(T1 m, T2 m.., AppMonad m) => m a -> m a` (they are almonst equivalent), but this constraint on b and functional deps makes it looking impossible or too hard.